File: LeetCode/longest_substring_without_repeating_characters.py
#!/usr/bin/python

class Solution:
    # @return an integer
    def search(self, s, l, r):
      count = [0] * 26
      i = l
      while i <= r:
        t = ord(s[i]) - ord('a')
        if count[t] == 0:
          count[t] += 1
        else:
          return False
        i += 1
      return True
    
    def lengthOfLongestSubstring(self, s):
      n = len(s)
      l = 1
      r = 27
      while l+1 < r:
        mid = (l+r) / 2
        i = 0
        flag = False
        while i + mid - 1 < n:
          if self.search(s, i, i+mid-1):
            flag = True
          i += 1
        if flag:
          l = mid
        else:
          r = mid
      return l


ss = Solution()
s = 'ityvnqitzvgcyqezmnpydkrsniczxiyidzdspolqjttkmzjaobyrnpka'
print ss.lengthOfLongestSubstring(s)

File: LeetCode/longest_substring_without_repeating_characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
      if (s.size() == 0) return 0;
      int ans;
      int i, j, l, r, n, vis[26];
      ans = 0;
      n = s.size();
      l = 1;
      r = n + 1;
      while (l + 1 < r) {
        int mid = (l+r) >> 1;
        int flag = 0;
        for (i = 0; i + mid - 1 < n; ++i) {
          memset(vis, 0, sizeof(vis));
          int ff = 1;
          for (j = i; j <= i + mid - 1; ++j)
            if (vis[s[j]-'a'] == 0)
              vis[s[j]-'a'] = 1;
            else {
              ff = 0;
              break;
            }
          if (ff) {
            flag = 1;
            break;
          }
        }
        if (flag) l = mid;
        else r = mid;
      }
      return l;
    }
};
